File: DataStructures.php
<?php

/* ------------------ARRAY------------------  */

// INITIALIZING
$slanguages = array();
$sdatabases = [];
$soperatingystems = array(12);

// POPULATING (SIMPLE)
$slanguages = [
    0 => 'Languages',
    1 => 'C++',
    2 => 'C',
    3 => 'Java',
    4 => 'C#',
    5 => 'Python',
    6 => 'PHP', 
    7 => 'Perl',
    8 => 'R',
    9 => 'Ruby',
    10 => 'VB.NET',
    11 => 'Javascript',
    12 => 'Matlab'];

$sdatabases[0] = 'Oracle';
$sdatabases[1] = 'SQL Server';
$sdatabases[2] = 'PostgreSQL';
$sdatabases[3] = 'MySQL';
$sdatabases[4] = 'DB2';
$sdatabases[5] = 'SQLite';
$sdatabases[6] = 'MS Access';
$sdatabases[7] = 'SAP Sybase';
$sdatabases[8] = 'Teradata';
$sdatabases[9] = 'MongoDB';

$soperatingystems = array('Windows 7', 'Windows XP', 'Windows 8',
                         'Windows 8.1', 'Windows 10', 'Windows Vista',
                         'Mac OS X 11', 'Mac OS X 10', 'Linux Mint',
                         'Linux Debian', 'Android', 'iOS');


// POPULATING (NESTED)
$nlanguages = [
    0 => ['Languages', 'Uses', 'Ranking'],
    1 => ['C++', 'computing', 99.6],
    2 => ['C', 'computing', 99.9],
    3 => ['Java', 'application', 100],
    4 => ['C#', 'application', 91.8],
    5 => ['Python', 'application', 95.8],
    6 => ['PHP',  'web', 84.5],
    7 => ['Perl', 'web', 66.9],
    8 => ['R', 'computing', 84.7],
    9 => ['Ruby', 'web', 75.3],
    10 => ['VB.NET', 'application', 63.4],
    11 => ['Javascript', 'web', 83],
    12 => ['Matlab', 'computing', 72.4]
];

echo "\nMULTI-DIMENSIONAL ARRAY NESTED FOREACH LOOP\n";
foreach($nlanguages as $item){
    foreach($item as $element){
        echo $element."\n";
    }          
}

$ndatabases[0] = ['Databases', 'Type', 'Size', 'Ranking'];
$ndatabases[1] = ['Oracle', 'Proprietary', 'Server', 1497.55];
$ndatabases[2] = ['SQL Server', 'Proprietary', 'Server', 1123.16];
$ndatabases[3] = ['PostgreSQL', 'Open-Source', 'Server', 280.09];
$ndatabases[4] = ['MySQL', 'Open-Source', 'Server', 1298.54];
$ndatabases[5] = ['DB2', 'Proprietary', 'Server', 196.13];
$ndatabases[6] = ['SQLite', 'Open-Source', 'File', 100.85];
$ndatabases[7] = ['MS Access', 'Proprietary', 'File', 140.21];
$ndatabases[8] = ['SAP Sybase', 'Proprietary', 'Server', 81.47];
$ndatabases[9] = ['Teradata', 'Proprietary', 'Server', 75.72];
$ndatabases[10] = ['MongoDB', 'Open-Source', 'Store', 301.39];

echo "\nMULTI-DIMENSIONAL ARRAY NESTED FOR LOOP\n";
for($i=0; $i < sizeof($ndatabases); $i++){
    for($j=0; $j < sizeof($ndatabases[0]); $j++){        
        echo $ndatabases[$i][$j]."\n";
    }          
}

$noperatingsystems = array(
    0 => array('OS', 'Type', 'Kernel', 'Market Share'),
    1 => array('Windows 7', 'desktop', 'NT', 56.11),
    2 => array('Windows XP', 'desktop', 'NT', 10.59),
    3 => array('Windows 8', 'desktop', 'NT', 2.88),
    4 => array('Windows 8.1', 'desktop', 'NT', 11.15),
    5 => array('Windows 10', 'desktop', 'NT', 9),
    6 => array('Windows Vista', 'desktop', 'NT', 0),
    7 => array('Mac OS X 11', 'desktop', 'Unix', 2.66),
    8 => array('Mac OS X 10', 'desktop', 'Unix', 2.45),
    9 => array('Linux Mint', 'desktop', 'Linux', 0),
    10 => array('Linux Debian', 'desktop', 'Linux', 0),
    11 => array('Android', 'mobile', 'Linux', 48.12),
    12 => array('iOS', 'mobile', 'Unix', 34.71)
);

echo "\nMULTI-DIMENSIONAL ARRAY NESTED WHILE LOOP\n";
$i = 0; $j = 0;
while($i < sizeof($noperatingsystems)){    
    while($j < sizeof($noperatingsystems[0])){        
        echo $noperatingsystems[$i][$j]."\n";
        $j++;
    }
    $i++; $j = 0;
}

/* DOUBLY LINK LIST  */

$dlanguages = new SplDoublyLinkedList();

$dlanguages->push(['Languages', 'Uses', 'Ranking']);
$dlanguages->push(['C++', 'computing', 99.6]);
$dlanguages->push(['C', 'computing', 99.9]);
$dlanguages->push(['Java', 'application', 100]);
$dlanguages->push(['C#', 'application', 91.8]);
$dlanguages->push(['Python', 'application', 95.8]);
$dlanguages->push(['PHP',  'web', 84.5]);
$dlanguages->push(['Perl', 'web', 66.9]);
$dlanguages->push(['R', 'computing', 84.7]);
$dlanguages->push(['Ruby', 'web', 75.3]);
$dlanguages->push(['VB.NET', 'application', 63.4]);
$dlanguages->push(['Javascript', 'web', 83]);
$dlanguages->push(['Matlab', 'computing', 72.4]);

echo "\nDOUBLY LINK LIST LOOP: FIFO\n";
$dlanguages->setIteratorMode(SplDoublyLinkedList::IT_MODE_FIFO);
for($dlanguages->rewind(); $dlanguages->valid(); $dlanguages->next()){
    echo $dlanguages->current()[0]."\n";
    echo $dlanguages->current()[1]."\n";
    echo $dlanguages->current()[2]."\n";       
}

/* FIXED ARRAY  */

$fdatabases = new SplFixedArray(11);

$fdatabases[0] = ['Databases', 'Type', 'Size', 'Ranking'];
$fdatabases[1] = ['Oracle', 'Proprietary', 'Server', 1497.55];
$fdatabases[2] = ['SQL Server', 'Proprietary', 'Server', 1123.16];
$fdatabases[3] = ['PostgreSQL', 'Open-Source', 'Server', 280.09];
$fdatabases[4] = ['MySQL', 'Open-Source', 'Server', 1298.54];
$fdatabases[5] = ['DB2', 'Proprietary', 'Server', 196.13];
$fdatabases[6] = ['SQLite', 'Open-Source', 'File', 100.85];
$fdatabases[7] = ['MS Access', 'Proprietary', 'File', 140.21];
$fdatabases[8] = ['SAP Sybase', 'Proprietary', 'Server', 81.47];
$fdatabases[9] = ['Teradata', 'Proprietary', 'Server', 75.72];
$fdatabases[10] = ['MongoDB', 'Open-Source', 'Store', 301.39];

echo "\nSPL FIXED ARRAY FOR LOOP\n";
for($i=0; $i<10; $i++){
    echo $fdatabases[$i][0]."\n";
    echo $fdatabases[$i][1]."\n";
    echo $fdatabases[$i][2]."\n";
    echo $fdatabases[$i][3]."\n";
}

/* SPL QUEUE  */

$qoperatingsystems = new SplQueue();

$qoperatingsystems->push(array('Windows 7', 'desktop', 'NT', 56.11));
$qoperatingsystems->push(array('Windows XP', 'desktop', 'NT', 10.59));
$qoperatingsystems->push(array('Windows 8', 'desktop', 'NT', 2.88));
$qoperatingsystems->push(array('Windows 8.1', 'desktop', 'NT', 11.15));
$qoperatingsystems->push(array('Windows 10', 'desktop', 'NT', 9));
$qoperatingsystems->push(array('Windows Vista', 'desktop', 'NT', 0));
$qoperatingsystems->push(array('Mac OS X 11', 'desktop', 'Unix', 2.66));
$qoperatingsystems->push(array('Mac OS X 10', 'desktop', 'Unix', 2.45));
$qoperatingsystems->push(array('Linux Mint', 'desktop', 'Linux', 0));
$qoperatingsystems->push(array('Linux Debian', 'desktop', 'Linux', 0));
$qoperatingsystems->push(array('Android', 'mobile', 'Linux', 48.12));
$qoperatingsystems->push(array('iOS', 'mobile', 'Unix', 34.71));

echo "\nSPL QUEUE ARRAY FOR LOOP: FIFO\n";
$qoperatingsystems->setIteratorMode(SplDoublyLinkedList::IT_MODE_FIFO);
for($qoperatingsystems->rewind(); $qoperatingsystems->valid(); $qoperatingsystems->next()){
    echo $qoperatingsystems->current()[0]."\n";
    echo $qoperatingsystems->current()[1]."\n";
    echo $qoperatingsystems->current()[2]."\n";
    echo $qoperatingsystems->current()[3]."\n"; 
}


/* ------------------BASIC OPERATIONS------------------ */

// SORT
// BY LANGUAGE RANKING
usort($nlanguages, function($a, $b) {
    return $b[2] - $a[2];
});

// FILTER
// BY DATABASE SIZE
function serveronly($var) {
    if ($var[2]=='Server'){
        return($var);
    }
}
$sdatabases = array_filter($ndatabases,"serveronly");

// AGGREGATE
// CALCULATE AVG. MARKET SHARE
$desktopavgshare = $mobileavgshare = 0;
foreach($noperatingsystems as $item){
    if ($item[1] == 'desktop') {
        $desktopavgshare = $desktopavgshare + $item[3];
    } else {
        $mobileavgshare = $mobileavgshare + $item[3];
    }
}
$desktopavgshare = $desktopavgshare/10;
$mobileavgshare = $mobileavgshare/2;

echo "DESKTOP AVG MARKET SHARE: ". $desktopavgshare."\n";
echo "MOBILE AVG MARKET SHARE: ". $mobileavgshare."\n";

// MERGE
$comparray = array('languages' => $nlanguages,
                   'databases' => $ndatabases,
                   'operatingsystems' => $noperatingsystems);


// CSV EXPORT
$cd = dirname(__FILE__);
$fp = fopen($cd.'/Data/Languages_PHP.csv', 'w');

fputcsv($fp, $nlanguages[12]);           # HEADERS
$nlanguages = array_slice($nlanguages, 0, 12);
foreach ($nlanguages as $rows) {    
    fputcsv($fp, $rows);                # ROWS
}
fclose($fp);

$fp = fopen($cd.'/Data/Databases_PHP.csv', 'w');
fputcsv($fp, $ndatabases[0]);           # HEADERS
$ndatabases = array_slice($ndatabases, 1, 10);
foreach ($ndatabases as $rows) {    
    fputcsv($fp, $rows);                # ROWS
}
fclose($fp);

$fp = fopen($cd.'/Data/OperatingSystems_PHP.csv', 'w');
fputcsv($fp, $noperatingsystems[0]);    # HEADERS
$noperatingsystems = array_slice($noperatingsystems, 1, 12);
foreach ($noperatingsystems as $rows) {    
    fputcsv($fp, $rows);                # ROWS
}
fclose($fp);

echo "\nSuccessfully exported array data to csv files!\n";

?>

File: DataStructures.py

######################################################
## LIST
##  CONTAINER OF OBJECTS MAPPED  TO VALUES
######################################################

'''DEFINING'''
slanguages = []
sdatabases = []
soperatingsystems = [None] * 12

'''POPULATING (SIMPLE)'''
slanguages = ['C++', 'C', 'Java', 'C#',
              'Python', 'PHP', 'Perl',
              'R', 'Ruby', 'VB.NET', 'Javascript', 'Matlab']

sdatabases.append('Oracle')
sdatabases.append('SQL Server')
sdatabases.append('PostgreSQL')
sdatabases.append('MySQL')
sdatabases.append('DB2')
sdatabases.append('SQLite')
sdatabases.append('MS Access')
sdatabases.append('SAP Sybase')
sdatabases.append('Teradata')
sdatabases.append('MongoDB')

soperatingsystems[0] = 'Windows 7'
soperatingsystems[1] = 'Windows XP'
soperatingsystems[2] = 'Windows 8'
soperatingsystems[3] = 'Windows 8.1'
soperatingsystems[4] = 'Windows 10'
soperatingsystems[5] = 'Windows Vista'
soperatingsystems[6] = 'Mac OS X 11'
soperatingsystems[7] = 'Mac OS X 10'
soperatingsystems[8] = 'Linux Mint'
soperatingsystems[9] = 'Linux Debian'
soperatingsystems[10] = 'Android'
soperatingsystems[11] = 'iOS'


'''POPULATING (NESTED)'''
nlanguages = []
ndatabases = []
noperatingsystems = list(range(13))

nlanguages = [['Languages', 'Uses', 'Ranking'],
              ['C++', 'computing', '99.6'],
              ['C', 'computing', '99.9'],
              ['Java', 'application', '100'],
              ['C#', 'application', '91.8'],
              ['Python', 'application', '95.8'],
              ['PHP', 'web', '84.5'],
              ['Perl', 'web', '66.9'],
              ['R', 'computing', '84.7'],
              ['Ruby', 'web', '75.3'],
              ['VB.NET', 'application', '63.4'],
              ['Javascript', 'web', '83'],
              ['Matlab', 'computing', '72.4']]

print('\nLANGUAGES LIST')
for item in nlanguages:
    print(item)

ndatabases.append(['Databases', 'Type', 'Size', 'Ranking'])
ndatabases.append(['Oracle', 'Proprietary', 'Server', '1497.55'])
ndatabases.append(['SQL Server', 'Proprietary', 'Server', '1123.16'])
ndatabases.append(['PostgreSQL', 'Open-Source', 'Server', '280.09'])
ndatabases.append(['MySQL', 'Open-Source', 'Server', '1298.54'])
ndatabases.append(['DB2', 'Proprietary', 'Server', '196.13'])
ndatabases.append(['SQLite', 'Open-Source', 'File', '100.85'])
ndatabases.append(['MS Access', 'Proprietary', 'File', '140.21'])
ndatabases.append(['SAP Sybase','Proprietary', 'Server', '81.47'])
ndatabases.append(['Teradata', 'Proprietary', 'Server', '75.72'])
ndatabases.append(['MongoDB', 'Open-Source', 'Server', '301.39'])

print('\nDATABASES LIST')
for item in ndatabases:
    print(item)
    
noperatingsystems[0] = ['Operating Systems', 'Type', 'Kernel', 'Market Share']
noperatingsystems[1] = ['Windows 7 ', 'desktop', 'NT', '56.11']
noperatingsystems[2] = ['Windows XP', 'desktop', 'NT', '10.59']
noperatingsystems[3] = ['Windows 8', 'desktop', 'NT', '2.88']
noperatingsystems[4] = ['Windows 8.1', 'desktop', 'NT', '11.15']
noperatingsystems[5] = ['Windows 10', 'desktop', 'NT', '9']
noperatingsystems[6] = ['Windows Vista', 'desktop', 'NT', '0']
noperatingsystems[7] = ['Mac OS X 11', 'desktop', 'Unix', '2.66']
noperatingsystems[8] = ['Mac OS X 10', 'desktop', 'Unix', '2.45']
noperatingsystems[9] = ['Linux Mint', 'desktop', 'Linux', '0']
noperatingsystems[10] = ['Linux Debian', 'desktop', 'Linux', '0']
noperatingsystems[11] = ['Android', 'mobile', 'Linux', '48.12']
noperatingsystems[12] = ['iOS', 'mobile', 'Unix', '34.71']

print('\nOPERATING SYSTEMS LIST')
for item in nlanguages:
    print(item)

######################################################
## DICTIONARY
##  MAPPING OBJECT INDEXING VALUES BY KEYS
######################################################

'''DEFINING'''
dlanguages = {}
ddatabases = {}
doperatingsystems = {}

'''POPULATING'''
dlanguages = {'Languages': ['Uses', 'Ranking'],
              'C++': ['computing', '99.6'],
              'C': ['computing', '99.9'],
              'Java': ['application', '100'],
              'C#': ['application', '91.8'],
              'Python': ['application', '95.8'],
              'PHP': ['web', '84.5'],
              'Perl': ['web', '66.9'],
              'R': ['computing', '84.7'],
              'Ruby': ['web', '75.3'],
              'VB.NET': ['application', '63.4'],
              'Javascript': ['web', '83'],
              'Matlab': ['computing', '72.4']}

print('\nLANGUAGES DICTIONARY')
for k, v in dlanguages.items():
    print('KEY: '+k)
    print(v)
    
ddatabases.update({'Databases': ['Type', 'Size', 'Ranking']})
ddatabases.update({'Oracle': ['Proprietary', 'Server', '1497.55']})
ddatabases.update({'SQL Server': ['Proprietary', 'Server', '1123.16']})
ddatabases.update({'PostgreSQL': ['Open-Source', 'Server', '280.09']})
ddatabases.update({'MySQL': ['Open-Source', 'Server', '1298.54']})
ddatabases.update({'DB2': ['Proprietary', 'Server', '196.13']})
ddatabases.update({'SQLite': ['Open-Source', 'File', '100.85']})
ddatabases.update({'MS Access': ['Proprietary', 'File', '140.21']})
ddatabases.update({'SAP Sybase': ['Proprietary', 'Server', '81.47']})
ddatabases.update({'Teradata': ['Proprietary', 'Server', '75.72']})
ddatabases.update({'MongoDB': ['Open-Source', 'Server', '301.39']})

print('\nDATABASES DICTIONARY')
for k, v in ddatabases.items():
    print('KEY: '+k)
    print(v)

doperatingsystems['Operating Systems'] = ['Type', 'Kernel', 'Market Share']
doperatingsystems['Windows 7 '] = ['desktop', 'NT', '56.11']
doperatingsystems['Windows XP'] = ['desktop', 'NT', '10.59']
doperatingsystems['Windows 8'] = ['desktop', 'NT', '2.88']
doperatingsystems['Windows 8.1'] = ['desktop', 'NT', '11.15']
doperatingsystems['Windows 10'] = ['desktop', 'NT', '9']
doperatingsystems['Windows Vista'] = ['desktop', 'NT', '0']
doperatingsystems['Mac OS X 11'] = ['desktop', 'Unix', '2.66']
doperatingsystems['Mac OS X 10'] = ['desktop', 'Unix', '2.45']
doperatingsystems['Linux Mint'] = ['desktop', 'Linux', '0']
doperatingsystems['Linux Debian'] = ['desktop', 'Linux', '0']
doperatingsystems['Android'] = ['mobile', 'Linux', '48.12']
doperatingsystems['iOS'] = ['mobile', 'Unix', '34.71']

print('\nOPERATING SYSTEMS DICTIONARY')
for k, v in doperatingsystems.items():
    print('KEY: '+k)
    print(v)


######################################################
## TUPLIE
##  IMMUTABLE CONTAINER OF OBJECTS MAPPED TO VALUES
######################################################

'''POPULATING (SIMPLE)'''
tlanguages = ('C++', 'C', 'Java', 'C#',
              'Python', 'PHP', 'Perl',
              'R', 'Ruby', 'VB.NET', 'Javascript', 'Matlab')

'''POPULATING (NESTED)'''
tdatabases = (('Databases', 'Type', 'Size', 'Ranking'),
              ('Oracle', 'Proprietary', 'Server', '1497.55'),
              ('SQL Server', 'Proprietary', 'Server', '1123.16'),
              ('PostgreSQL', 'Open-Source', 'Server', '280.09'),
              ('MySQL', 'Open-Source', 'Server', '1298.54'),
              ('DB2', 'Proprietary', 'Server', '196.13'),
              ('SQLite', 'Open-Source', 'File', '100.85'),
              ('MS Access', 'Proprietary', 'File', '140.21'),
              ('SAP Sybase','Proprietary', 'Server', '81.47'),
              ('Teradata', 'Proprietary', 'Server', '75.72'),
              ('MongoDB', 'Open-Source', 'Server', '301.39'))

toperatingsystems= {'Windows 7 ': ('desktop', 'NT', '56.11'),
                     'Windows XP': ('desktop', 'NT', '10.59'),
                     'Windows 8': ('desktop', 'NT', '2.88'),
                     'Windows 8.1': ('desktop', 'NT', '11.15'),
                     'Windows 10': ('desktop', 'NT', '9'),
                     'Windows Vista': ('desktop', 'NT', '0'),
                     'Mac OS X 11': ('desktop', 'Unix', '2.66'),
                     'Mac OS X 10': ('desktop', 'Unix', '2.45'),
                     'Linux Mint': ('desktop', 'Linux', '0'),
                     'Linux Debian': ('desktop', 'Linux', '0'),
                     'Android': ('mobile', 'Linux', '48.12'),
                     'iOS': ('mobile', 'Unix', '34.71')}


######################################################
## SET
##  UNORDERED COLLECTION OF UNIQUE ITEMS
######################################################

stlanguages = set(slanguages)
print("\nLANGUAGES SET")
print(stlanguages)

stdatabases = set(ddatabases)
print("\nDATABASES SET")
print(stdatabases)

stoperatingsystems = set(toperatingsystems)
print("\nOPERATING SYSTEMS SET")
print(stoperatingsystems)


############################################################
## MATRIX AND ARRAY (NUMPY MODULE)
##  TWO-DIMENSIONAL/N-DIMENSIONAL ROW BY COLUMN STRUCTURES
############################################################

import numpy as np

mxlanguages = np.matrix([['Languages', 'Uses', 'Ranking'],
                         ['C++', 'computing', '99.6'],
                         ['C', 'computing', '99.9'],
                         ['Java', 'application', '100'],
                         ['C#', 'application', '91.8'],
                         ['Python', 'application', '95.8'],
                         ['PHP', 'web', '84.5'],
                         ['Perl', 'web', '66.9'],
                         ['R', 'computing', '84.7'],
                         ['Ruby', 'web', '75.3'],
                         ['VB.NET', 'application', '63.4'],
                         ['Javascript', 'web', '83'],
                         ['Matlab', 'computing', '72.4']], dtype='str')
print('\nLANGUAGES NUMPY MATRIX')
print(mxlanguages)

mxdatabases = np.matrix(ndatabases, dtype='str')
print('\nDATABASES NUMPY MATRIX')
print(mxdatabases)
      
arroperatingsystems = np.array(noperatingsystems, dtype='str')
print('\nOPERATING SYSTEMS NUMPY ARRAY')
print(arroperatingsystems)

############################################################
##  BASIC OPERATIONS
############################################################

## SORT
slanguages.sort()

def getKey(item):
    num = 99999999 if item[3] == 'Ranking' else float(item[3])
    return num
ndatabases.sort(key=getKey, reverse=True)

## FILTER
soperatingsystems = [x if 'Windows' in x else '' for x in soperatingsystems]

noperatingsystems = [noperatingsystems[i] if 'Windows' in noperatingsystems[i][0] \
                                          else ['']*4 for i in range(len(noperatingsystems))]
noperatingsystems = list(filter(lambda x: x != ['']*4, noperatingsystems))

## AGGREGATE
avgrank = 0
for i in range(1,len(mxlanguages)):    
    avgrank += np.sum(float(mxlanguages[i,2]))

print("AVERAGE LANGUAGE RANK: {}".format(round(avgrank/12),3))


desktopavgmktshare = mobileavgmktshare = 0
desktopcount = mobilecount = 0
for os in toperatingsystems:
    
    if toperatingsystems.get(os)[0] == 'desktop' and toperatingsystems.get(os)[2] != 'Ranking':
        desktopavgmktshare = desktopavgmktshare + round(float(toperatingsystems.get(os)[2]),3)
        desktopcount += 1
        
    if toperatingsystems.get(os)[0] == 'mobile' and toperatingsystems.get(os)[2] != 'Ranking':
        mobileavgmktshare = mobileavgmktshare + round(float(toperatingsystems.get(os)[2]),3)
        mobilecount += 1

print("\nDESKTOP AVG. MARKET SHARE: {}".format(desktopavgmktshare/desktopcount))
print("\nMOBILE AVG. MARKET SHARE: {}".format(mobileavgmktshare/mobilecount))

## MERGE
complist = [nlanguages, ndatabases, noperatingsystems]

compdict = {'langauges': {k:v for (k,v) in dlanguages.items() if not 'Languages' in k},
            'databases': {k:v for (k,v) in ddatabases.items() if not 'Databases' in k},
            'operatingsystems': {k:v for (k,v) in doperatingsystems.items() if not 'Operating Systems' in k}}

comparray = np.array([np.matrix(nlanguages, dtype='str'),
                      np.matrix(ndatabases, dtype='str'),
                      np.matrix(noperatingsystems, dtype='str')])
    

############################################################
## DATA FRAME (PANDAS MODULE)
##  TWO-DIMENSIONAL DATASET STRUCTURE
############################################################

import os
import pandas as pd

cd = os.path.dirname(os.path.abspath(__file__))

langdf = pd.DataFrame(nlanguages)
langdf.columns = nlanguages[0]
langdf = langdf[1:]
      
print('\nLANGUAGES PANDAS DATA FRAME')
print(langdf)
langdf.to_csv(os.path.join(cd, 'Data', 'Languages_PY.csv'), index=False)

dbdf = pd.DataFrame(ndatabases)
dbdf.columns = ndatabases[0]
dbdf = dbdf[1:]

print('\nDATABASES PANDAS DATA FRAME')
print(dbdf)
dbdf.to_csv(os.path.join(cd, 'Data', 'Databases_PY.csv'), index=False)

osdf = pd.DataFrame(toperatingsystems)
osdf = osdf.transpose()
osdf = osdf.reset_index()
osdf.columns = ['OS', 'Type', 'Kernel', 'Market Share']

print('\nOPERATING SYSTEMS PANDAS DATA FRAME')
print(osdf)
osdf.to_csv(os.path.join(cd, 'Data', 'OperatingSystems_PY.csv'), index=False)

print('\nSuccessfully outputted dataframes to csv!')
